src/server/pipeline.py

```diff
@@ -209,8 +214,15 @@ async def perform_handshake(
             raise Exception("No longer accepting handshakes, closing.")
 
         if not global_connections.add(connection):
+            # Don't penalize a duplicate connection.
+            mark_attempted = False
```

Comment:
I don't get how this can do anything when you immediately raise in the next line.